### ISSUE TITLE
[Feature request] direnv (directory specific env variables)

### ISSUE THREAD

**Issue by c33s** [cmderdev/cmder#2769]
### Suggestion

as https://direnv.net/ is not 100% windows compatible it would be nice to have its function integrated in cmder. direnv is an extension for your shell. It augments existing shells with a new feature that can load and unload environment. it checks if a file `.envrc` or `.env` exists in the current directory and loads/unloads the given env vars for this directory before executing anything else

### Use case

- Load 12factor apps environment variables
- Create per-project isolated development environments
- Load secrets for deployment

### Checklist

- [X] I have read the documentation and made sure this feature doesn't already exist.
- [X] I have searched for similar issues and found none that describe my feature request.

**Comment by chrisant996** [comment]
Some observations, for future reference:

direnv is a lot more complicated than just a Lua script.

It allows running .envrc scripts in a separate bash shell to determine environment variable changes and then apply them back to the original shell (which might not even be bash).  And it provides a library of functions for use in the .envrc files, which can be invoked via `direnv functionname args`, so a Lua script isn't enough, there would need to be a whole separate program as well.

A simplistic version could be made that has much less functionality, but to actually support `.env` and `.envrc` files on Windows is more complicated than it might sound at first.  I haven't yet explored how direnv supports powershell -- maybe there are `.env.ps1` files or something.

**Comment by chrisant996** [comment]
Right.  A `.env.cmd` file would be feasible, but would not work in a cross-platform way.

With that said, I have a prototype of a `direnv.lua` script that I'm experimenting with, and which I may include in clink-gizmos at some point. 

I've had a private script that does something similar, but had hard-coded directories and envvars.  Extending that to be configurable isn't a big stretch, and benefits me directly, so at some point that'll show up in clink-gizmos.

**Comment by DRSDavidSoft** [comment]
Wow, that would be amazing! A huge plus point to extending private scripts to be included in clink-gizmos where it would benefit anyone with similar use cases.

In any case, we should also point users to clink-gizmos more as it includes some excellent tools. I believe it is also a great fit to be bundled with Cmder, in addition to the clink completions. Probably need to open a PR soon for this.

Note that while `.env.cmd` is a good point, I believe the most appealing use case for direnv would be due to its cross-platform-ness, so my vote is the `.env` file, which is a simple key pair list.

If so, adding this to clink gizmos in my opinion will check the box for clink shell, if you also agree.

As for PowerShell, I have some ideas in mind, but I'd prefer to base the behavior on what the primary shell of Cmder, which is clink, does.

**Comment by DRSDavidSoft** [comment]
Amazing, thanks @chrisant996! 🚀

**Comment by DRSDavidSoft** [comment]
Quick question, suppose there is an env var already set. If we navigate to a directory that contains a `.env` file with a diff then exit that directory, will the env var reset to the previous value, or will it be unset?

Also, will the env vars be used for any subdirectories that we `cd` into, or just the directory that contains the `.env` file?

I would appreciate it if you could provide a summary of how the clink gizmo's implementation works, as the documentations didn't describe it in detail.

Nice work security-wise (being disabled by default, and having to explicitly trust directories). 👍🏻 Top notch work!

**Comment by chrisant996** [comment]
Direnv is apparently available for Windows, it just isn't present by default and doesn't integrate with CMD (but it does supposedly have a way to integrate with PowerShell).

I'm going to explore whether it's better to make direnv.lua simply integrate with the original direnv program from the direnv.net project, for optimal consistency and interoperability.

From Cmder's perspective, if the original direnv program can integrate with Clink through a custom Lua script, then that would enable direnv to work the same across all shells supported by Cmder.

**Comment by chrisant996** [comment]
An example of not working on Windows:

```console
> direnv.exe exec c:\windows\system32\cmd.exe /c set CLINK
direnv: error command 'c:\windows\system32\cmd.exe' not found on PATH ''
```

The "exec" capability in direnv.net seems to have several problems on Windows (the example above illustrates just one of the problems).  Furthermore, the PATH envvar is not empty, and yet the direnv.net project's implementation can't see the PATH envvar, and it shouldn't be looking at PATH anyway because a fully qualified pathname was provided.
